src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -1342,34 +1343,45 @@ void ZooKeeper::receiveEvent()
 
 void ZooKeeper::finalize(bool error_send, bool error_receive)
 {
+    bool check = false;
+    /// If some thread (send/receive) already finalizing session don't try to do it
+    if (!finalization_started.compare_exchange_strong(check, true))
```

Comment:
I think `exchange` will be enough, because it returns previous value.

Reply:
Much simplier, thanks